interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp

```diff
@@ -433,29 +507,99 @@ namespace cling {
-             DirIt != DirEnd && !EC; DirIt.increment(EC)) {
-
-          // FIXME: Use a StringRef here!
-          std::string FileName = getRealPath(DirIt->path());
-          assert(!llvm::sys::fs::is_symlink_file(FileName));
-
+        std::function<const LibraryPath*(std::string, unsigned)> HandleLib =
+            [&](std::string FileName, unsigned level) -> const LibraryPath* {
           if (ShouldPermanentlyIgnore(FileName))
-            continue;
+            return nullptr;
 
           std::string FileRealPath = llvm::sys::path::parent_path(FileName);
           FileName = llvm::sys::path::filename(FileName);
           const BasePath& BaseP = m_BasePaths.RegisterBasePath(FileRealPath);
           LibraryPath LibPath(BaseP, FileName);
-          if (m_SysLibraries.HasRegisteredLib(LibPath) ||
-              m_Libraries.HasRegisteredLib(LibPath))
-            continue;
 
+          if (auto result = m_SysLibraries.GetRegisteredLib(LibPath))
+            return &(*result);
+          if (auto result = m_Libraries.GetRegisteredLib(LibPath))
+            return &(*result);
+
+          const LibraryPath* NewLibPath;
           if (searchSystemLibraries)
-            m_SysLibraries.RegisterLib(LibPath);
+            NewLibPath = m_SysLibraries.RegisterLib(LibPath);
           else
-            m_Libraries.RegisterLib(LibPath);
+            NewLibPath = m_Libraries.RegisterLib(LibPath);
+
+          // Handle lib dependencies
+          std::string RPath;
+          std::string RunPath;
+          std::vector<StringRef> Deps;
+          llvm::StringRef NewLibP = NewLibPath->GetFullName();
+          auto ObjFileOrErr =
+              llvm::object::ObjectFile::createObjectFile(NewLibP);
+          if (llvm::Error Err = ObjFileOrErr.takeError()) {
+            if (DEBUG > 1) {
+              std::string Message;
+              handleAllErrors(std::move(Err), [&](llvm::ErrorInfoBase& EIB) {
+                Message += EIB.message() + "; ";
+              });
+              cling::errs()
+                  << "Dyld::ScanForLibraries: Failed to read object file "
+                  << NewLibP << " Errors: " << Message << "\n";
+            }
+            return nullptr;
+          }
+          llvm::object::ObjectFile* BinObjF = ObjFileOrErr.get().getBinary();
+          if (BinObjF->isELF()) {
+            if (const auto* ELF = dyn_cast<ELF32LEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
+            else if (const auto* ELF = dyn_cast<ELF32BEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
+            else if (const auto* ELF = dyn_cast<ELF64LEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
+            else if (const auto* ELF = dyn_cast<ELF64BEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
```

Comment:
We shouldn't be looking at non-process ABI: if this process is 64bit, only look at 64bit ELF; if it's 32 only look at 32bit ELF. Same for endianness. Or what am I missing?

Reply:
We already seem to have that filtered by https://github.com/vgvassilev/cling/blob/f3768a4c43b0f3b23eccc6075fa178861a002a10/lib/Interpreter/DynamicLibraryManagerSymbol.cpp#L730-L731